advance georefence street address

matallo opened the pull request:
re: #11662 https://github.com/CartoDB/cartodb/issues/11812

many changes here:

- added Toggle editor (based in Radius)
- added Placeholder add-on to Codemirror
- added `autocompleteChars`, `autocompleteSuffix`, `autocompletePrefix `, `autocompleteTriggers` (these two come usually together) options to Codemirror
- added mustache mode to Codemirror
- extend custom Textarea editor from custom Text editor
- added `editor` option to custom Textarea editor, which will turn it into a Codemirror editor passing the option `editor: true` this won't have any effect for current Textarea editors
- added placeholder option to Textarea custom editor
- made `querySchemaModel`, and `layerDefinitionModel` optional for hints
- added and extended the necessary tests for the changes above

**acceptance**

- current codemirror hints should keep working as usual
- new textarea codemirror should appear when selecting advance in the toggle of georeference analysis
- new textarea codemirror hints should appear
- normal georeference analysis should work (d'oh!)
- advance georeference analysis should work (d'oh!)

how to test the codemirror hints:

after writing "{{", or first two characters of columns, it should suggest column names, and autocomplete "}}" suffix, as well as "{{" prefix if it has not been typed

codemirror hints should work on demand with "Ctrl + Space" shortcut, with the same behaviour

they should work normally in previous codemirror hints such sql in data tab

**STAGING**: rui servers

xavijam commented:
cc @noguerol ^^^

urbanphes commented:
Some things;

- Less left and right margin interior on the darker input. We can test with 4px to each instead of 14px.
- Change the info by: “Use columns or free text to compose your address schema”
- Change the text color to difference the text inside {{}} and the text outside. Maybe we can use the white for the text outside like a “{“
- The autocomplete should appear when you start typing columns even if you don’t put keys.
- When there is an error, the input doesn't mark as red.

matallo commented:
- [x] Less left and right margin interior on the darker input. We can test with 4px to each instead of 14px.
- [x] Change the info by: “Use columns or free text to compose your address schema”
- [ ] Change the text color to difference the text inside {{}} and the text outside. Maybe we can use the white for the text outside like a “{“
- [x] The autocomplete should appear when you start typing columns even if you don’t put keys.
- [x] When there is an error, the input doesn't mark as red.

Changes made, the only pending point is the syntax color which I will experiment a bit, as we need to make our custom syntax coloring, this is deployed again in the ruis, and tests are passing 🤞

saleiva commented:
@matallo @urbanphes @noguerol should we do another pass then? The point that is missing could be a nice-to-have at some point...

matallo commented:
it's ready to test again 👍

I'm doing some quick experiment now to see if I can get the syntax coloring working, in any case

xavijam commented on lib/assets/core/javascripts/cartodb3/components/code-mirror/code-mirror-view.js:
```diff
@@ -27,16 +29,22 @@ module.exports = CoreView.extend({
     if (!opts.model) throw new Error('Model for codemirror is required.');
-    if (opts.model.get('content') === undefined) throw new Error('Content property for codemirror is required.');
-    if (!opts.tip) throw new Error('tip message is required');
+    if (opts.model.get('content') === void 0 &&
```
I'd use the nice and great `var checkAndBuildOpts = require('../../helpers/required-opts');` here. What do you think?

matallo replied:
this is a tricky one as we have some required opts indeed as the `model` but some other cases such either `content` or `placeholder` is required, and finally some optional opts like prefix/suffix, ...

I'm not really sure, thoughts @nobuti ?

nobuti replied:
Agree, we are mixing options with attributes, I don't think we can use the helper for all.

xavijam replied:
_flojos_

matallo replied:
haha, actually in this PR I replaced the `required_opts` for `optional_opts` (as redundant as it might seem) https://github.com/CartoDB/cartodb/pull/11939/files#diff-c262f4f35ff0590e20b9390bbc8316e3R6

so maybe it's the time to set a proper name for this where we can have both the required and the optional _whatever_ with a different loop for each one

xavijam commented on lib/assets/core/javascripts/cartodb3/components/code-mirror/code-mirror-view.js:
```diff
@@ -95,44 +115,92 @@ module.exports = CoreView.extend({
+  },
+
+  _completeIfAfterCtrlSpace: function (cm) {
+    var self = this;
```
Why do we need the `self` here?

matallo replied:
👍

xavijam commented on lib/assets/core/javascripts/cartodb3/components/code-mirror/code-mirror-view.js:
```diff
@@ -27,16 +29,22 @@ module.exports = CoreView.extend({
+        opts.placeholder === void 0) throw new Error('Content property or placeholder for codemirror is required.');
+    if (opts.tip === void 0) throw new Error('tip message is required');
 
+    this._autocompleteChars = opts.autocompleteChars || 3;
```
This 3 number should be a "static" var at the top. But in any case, if this is a view, you could create the `options` variable inside the Class and specify whatever you want (same for mode).

matallo replied:
👍 also moved the default attributes for the model in that `options:`

xavijam commented on lib/assets/core/javascripts/cartodb3/components/code-mirror/code-mirror-view.js:
```diff
@@ -95,44 +115,92 @@ module.exports = CoreView.extend({
 
-  _showAutocomplete: function (cm) {
+  _completeAfter: function (cm, pred, opts) {
+    var self = this;
```
Same as previously said with `self`.

matallo replied:
👍

xavijam commented on lib/assets/core/javascripts/cartodb3/components/form-components/editors/radio.tpl:
```diff
@@ -3,7 +3,7 @@
     name="<%- item.name %>" value="<%- item.value %>" id="<%- item.id %>" 
         <% if (item.selected) { %>
-          selected="selected"
+          checked="checked"
```
Won't it make fail or work weird for the rest of the Radio components?

matallo replied:
from MDN https://developer.mozilla.org/en/docs/Web/HTML/Element/input

> Use the checked attribute to indicate whether this item is selected by default

checked (pun intended) in several other places and it doesn't seem to have further consequences

xavijam replied:
😱

xavijam commented on lib/assets/core/javascripts/cartodb3/components/form-components/editors/textarea.js:
```diff
@@ -1,15 +1,38 @@
 
   initialize: function (opts) {
     Backbone.Form.editors.Base.prototype._setOptions.call(this, opts); // Options
-    this.constructor.__super__.initialize.apply(this, arguments);
```
Hey, I feel we should not remove this line, should we?

matallo replied:
now we're extending from custom `Text` editor instead of Backbone forms `Textarea` editor, where explicitly prevents this https://github.com/powmedia/backbone-forms/blob/master/src/editors/textarea.js#L9

xavijam replied:
Well, not sure about that, but you have tested it so 👍 .

xavijam commented on lib/assets/core/javascripts/cartodb3/components/form-components/editors/textarea.js:
```diff
@@ -1,15 +1,38 @@
   render: function () {
     this.setValue(this.value);
+
+    if (this.options.editor) {
```
Hum... 🤔  about if we should add this behaviour to the default textarea in the Backbone.Forms...

matallo replied:
how to better proceed? Having a new editor extending from `Textarea`? How that new editor will be called?

xavijam replied:
I'm not sure, it depends the number of uses. Will we need more features for this textarea? or more types? I feel we could leave it as it is for the moment, but if it grows, create it in a separate it editor.

nobuti replied:
Same concerns, this is an exception, and as it is, I think it should be a new editor. We are adding a lot of code that the majority of instances don't need.

matallo replied:
same concerns than @xavijam then, happy to move it out to its own editor 😄

xavijam replied:
🎉

saleiva commented:
taking it then

xavijam commented:
Hey @saleiva, not sure if it is deployed in any staging with last changes, @matallo?

matallo commented:
it is!

saleiva commented:
Works nice but I found a few things:
- [ ] Syntax errors are kept. Try with {{calle}, When applying it removes some text and do the analysis. We should validate on the back and give an error, or keep the text as it is.
- [ ] After applying the analysis, if you go to the analysis and modify only the text on the dark box, the button doesn't get activated back. 
- [ ] If changing the text on advanced, clicking back on normal mode, and going back to advanced, the template is not being saved.
- [ ] If being in the Street Address form, and clicking on the dropdown for selecting Admin regions, the form doesn't get updated.

matallo commented:
- [ ]  Syntax errors are kept. Try with {{calle}, When applying it removes some text and do the analysis. We should validate on the back and give an error, or keep the text as it is.

I'm not sure how to proceed with the syntax error :/ anyway, there was a bug and now the string is being sent as is, not removed 👍 

- [x] After applying the analysis, if you go to the analysis and modify only the text on the dark box, the button doesn't get activated back.

fixed some stuff that prevented changes in the codemirror to be populated to the form

- [x] If being in the Street Address form, and clicking on the dropdown for selecting Admin regions, the form doesn't get updated.

🙈 well spotted, fixed now

- [ ] If changing the text on advanced, clicking back on normal mode, and going back to advanced, the template is not being saved.

this one is not easy, but I can further look into it, I can share which is the process now, though: when changing context I'm copying the value in the `street_address_column` (select) to `street_address_template` (textarea) adding the wrap `{{*}}`, if I change back as the values won't coincide the select has not any option selected, and if I change back again, as the value is copied, the value copied is empty

as said, I'll take a look if I can workaround this somehow

matallo commented:
after discussing with buti, we cannot see a straight-forward solution for keeping the `street_address_template` value, also none of the forms work this way now, i.e.:

1. select the columns for longlat
1. change type
1. go back to longlat

the columns are not preserved

saleiva commented:
@matallo ok, I'd keep that as an enhancement for the time being
cc/ @noguerol

saleiva commented:
About the first point, can we validate the syntax of the expression? I talked to @rochoa about that

matallo commented:
Yeah, we'd probably have to think about it globally, for every form.

On Apr 11, 2017 4:58 PM, "Sergio Álvarez Leiva" <notifications@github.com>
wrote:

> @matallo <https://github.com/matallo> ok, I'd keep that as an enhancement
> for the time being
> cc/ @noguerol <https://github.com/noguerol>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CartoDB/cartodb/pull/11939#issuecomment-293290570>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACPRO3Z4ASIg6C7hnqCgmr9glXXBIIuks5ru5TsgaJpZM4M2u8Z>
> .
>

rochoa commented:
@saleiva, @matallo 

> About the first point, can we validate the syntax of the expression? I talked to @rochoa about that

The backend will keep the string as it is when the template is something like `{{calle} {{number}}, {{city}}`, replacing `{{number}}` and `{{city}}`, but keeping `{{calle}` like it is.

nobuti commented on app/assets/stylesheets/editor-3/codemirror.css.scss:
```diff
@@ -465,6 +465,19 @@ span.CodeMirror-selectedtext { background: none; }
+  color: #e9eded;
+}
+
+.cm-mustache-error {
```
I think we are not using this anymore.

matallo replied:
we are! we should get rid of it if we go ahead with https://github.com/CartoDB/cartodb/pull/11965 though

nobuti commented on lib/assets/core/javascripts/cartodb3/components/code-mirror/code-mirror-view.js:
```diff
@@ -95,44 +120,93 @@ module.exports = CoreView.extend({
+    var opts = {};
+    var cur = cm.getCursor();
+
+    var openingMustache = cm.getRange(CodeMirror.Pos(cur.line, cur.ch - 2), cur) !== '{{';
```
I feel this should not be here, it should be an option. The base view should not know anything about mustache or any other mode the user is using.

nobuti commented on lib/assets/core/javascripts/cartodb3/components/form-components/editors/toggle.js:
```diff
@@ -0,0 +1,36 @@
+var Backbone = require('backbone');
+var template = require('./toggle.tpl');
+
+Backbone.Form.editors.Toggle = Backbone.Form.editors.Radio.extend({
```
Sweet.

matallo commented:
back to development again, for someone to take those changes, thanks for the review mates!

matallo commented:
what about now @xavijam @nobuti ?

xavijam commented:
At a first sight looks what we said so 👍 , but in any case, when you have some minutes @nobuti, could you take a look please? 💋

nobuti commented:
👍

saleiva commented:
After another pass I found the following issues:
- [ ] After applying the analysis, the name of the type is weird on the selector (looks like it's not internacionalized) and it only appears the option to switch to lat/lon and not all.
<img width="308" alt="screen shot 2017-04-18 at 15 36 11" src="https://cloud.githubusercontent.com/assets/132149/25133384/cfc2c54e-244c-11e7-9915-34bef896457a.png">
<img width="1151" alt="screen shot 2017-04-18 at 15 36 49" src="https://cloud.githubusercontent.com/assets/132149/25133397/dfd5d7be-244c-11e7-80a3-81d575b2d02a.png">

- [ ] After applying the analysis I cannot apply it again. The button is always disabled.

I tested with [Relacion.Camaras.csv.zip](https://github.com/CartoDB/cartodb/files/929820/Relacion.Camaras.csv.zip)

matallo commented:
>  After applying the analysis, the name of the type is weird on the selector (looks like it's not internacionalized) and it only appears the option to switch to lat/lon and not all.

This is already happening in production :( investigating how to fix it, and pushing it to the PR

matallo commented:
thank you @nobuti very much, for maanging the `dsReady` issue, I just updated the syntax color for latin characters, anything else?

(still in the ruis)

xavijam commented:
@nobuti has found something weird with the behaviour of the "apply" button but it doesn't have to do with this PR, so :+1:. To acceptance again?

matallo commented:
promoted!

matallo commented:
and back to development, with the last changes there seems to be some condition by which applying a georeference analysis > subsample analysis > edit georeference analysis the subsample analysis gets stuck `enqueued` 

this is not happening in production but in staging

xavijam commented:
How come? I mean, how is that possible? JS error? take into account node states come from CARTO.js.

matallo commented:
no errors are shown in the console, I'm investigating further

matallo commented:
dismiss my last comment this seems to be an issue with staging environments rather than georeference
